python/gstaichi/lang/func.py

```diff
@@ -0,0 +1,173 @@
+            current_args_key
+        )
+        tree, ctx = self.get_tree_and_ctx(
+            is_kernel=False,
+            args=args,
+            ast_builder=self.current_kernel.ast_builder(),
+            is_real_function=self.is_real_function,
+            used_py_dataclass_parameters_enforcing=used_by_dataclass_parameters_enforcing,
+        )
+
+        struct_locals = _kernel_impl_dataclass.extract_struct_locals_from_context(ctx)
+
+        tree = _kernel_impl_dataclass.unpack_ast_struct_expressions(tree, struct_locals=struct_locals)
+        ret = transform_tree(tree, ctx)
+        self.current_kernel = None
+        if not self.is_real_function:
+            if self.return_type and ctx.returned != ReturnStatus.ReturnedValue:
+                raise GsTaichiSyntaxError("Function has a return type but does not have a return statement")
+        return ret
+
+    def func_call_rvalue(self, key: FunctionKey, args: tuple[Any, ...]) -> Any:
+        # Skip the template args, e.g., |self|
+        assert self.is_real_function
+        non_template_args = []
+        dbg_info = _ti_core.DebugInfo(impl.get_runtime().get_current_src_info())
+        for i, kernel_arg in enumerate(self.arg_metas):
+            anno = kernel_arg.annotation
+            if not isinstance(anno, template):
+                if id(anno) in primitive_types.type_ids:
+                    non_template_args.append(ops.cast(args[i], anno))
+                elif isinstance(anno, primitive_types.RefType):
+                    non_template_args.append(_ti_core.make_reference(args[i].ptr, dbg_info))
+                elif isinstance(anno, ndarray_type.NdarrayType):
+                    if not isinstance(args[i], AnyArray):
+                        raise GsTaichiTypeError(
+                            f"Expected ndarray in the kernel argument for argument {kernel_arg.name}, got {args[i]}"
+                        )
+                    non_template_args += _ti_core.get_external_tensor_real_func_args(args[i].ptr, dbg_info)
+                else:
+                    non_template_args.append(args[i])
+        non_template_args = impl.make_expr_group(non_template_args)
+        compiling_callable = impl.get_runtime().compiling_callable
+        assert compiling_callable is not None
+        func_call = compiling_callable.ast_builder().insert_func_call(
+            self.cxx_function_by_id[key.instance_id], non_template_args, dbg_info
+        )
+        if self.return_type is None:
+            return None
+        func_call = Expr(func_call)
+        ret = []
+
+        for i, return_type in enumerate(self.return_type):
+            if id(return_type) in primitive_types.type_ids:
+                ret.append(
+                    Expr(
+                        _ti_core.make_get_element_expr(
+                            func_call.ptr, (i,), _ti_core.DebugInfo(impl.get_runtime().get_current_src_info())
```

Comment:
You already computed `_ti_core.DebugInfo(impl.get_runtime().get_current_src_info())`